processor/src/continuation_stack.rs

```diff
@@ -0,0 +1,193 @@
+use alloc::{sync::Arc, vec::Vec};
+
+use miden_core::{
+    Program,
+    mast::{MastForest, MastNodeId},
+};
+
+/// Specifies the phase of a processing step.
+///
+/// Specifically, it allows to distinguish between the start and finish phases of processing a node,
+/// which gives the processor sufficient information to execute the node correctly. Note that
+/// exactly how "start" and "end" are used depends on the node type being processed.
+#[derive(Debug, Default)]
+pub enum ProcessingStepPhase {
```

Comment:
IMO this can just be combined with `ProcessingStep`, and then the prologue/epilogue of each node type handler be split out into separate functions.

This will be much friendlier to the branch predictor, because the dispatch loop will only call a given handler for a specific `ProcessingStep` variant, which in turn will lead to more predictable control flow patterns, i.e. every time the `execute_join_node_start` function is called from the call site in the match arm of `ProcessingStep::Prologue`, the branch predictor can predict with great certainty that the next branch in the dispatch loop will be to the `ProcessingStep::Node` arm and from there to `execute_join_node` - and so on.

Reply:
Fixed